raiden-ts/src/services/epics.ts

```diff
@@ -695,3 +706,153 @@ export const monitorRequestEpic = (
+        concatMap((txHash) =>
+          combineLatest([
+            from(provider.getTransactionReceipt(txHash)),
+            state$.pipe(pluckDistinct('blockNumber')),
+            config$.pipe(pluckDistinct('confirmationBlocks')),
+          ]).pipe(
+            first(
+              ([receipt, blockNumber, confirmationBlocks]) =>
+                !!receipt.blockNumber && receipt.blockNumber + confirmationBlocks <= blockNumber,
+            ),
+            map(() => txHash),
+          ),
+        ),
+        concatMap((txHash) => {
+          return from(userDepositContract.functions.withdraw_plans(address)).pipe(
+            map(({ amount, withdraw_block }) =>
+              udcWithdrawPlanned(
+                {
+                  txHash,
+                  block: withdraw_block.toNumber(),
+                },
+                {
+                  amount: amount as UInt<32>,
+                },
+              ),
+            ),
+          );
+        }),
+        catchError((err) => {
+          log.error('Planning udc withdraw failed', err);
+          return of(udcWithdraw.failure(err, action.meta));
+        }),
+      );
+    }),
+  );
+};
+
+export const udcCheckWithdrawPlannedEpic = (
+  action$: Observable<RaidenAction>,
+  state$: Observable<RaidenState>,
+  { userDepositContract, address }: RaidenEpicDeps,
+): Observable<udcWithdrawPlanned> => {
+  return action$.pipe(
+    filter(newBlock.is),
+    first(),
+    concatMap(() =>
+      from(userDepositContract.functions.withdraw_plans(address)).pipe(
+        filter((value) => value.withdraw_block.gt(Zero)),
+        map(({ amount, withdraw_block }) =>
+          udcWithdrawPlanned(
+            {
+              block: withdraw_block.toNumber(),
+            },
+            {
+              amount: amount as UInt<32>,
+            },
+          ),
+        ),
+      ),
+    ),
+  );
+};
+
+export const udcWithdrawPlannedEpic = (
+  action$: Observable<RaidenAction>,
+  state$: Observable<RaidenState>,
+  { log, userDepositContract, address, signer }: RaidenEpicDeps,
+): Observable<udcWithdraw.success | udcWithdraw.failure> => {
+  return action$.pipe(filter(udcWithdrawPlanned.is)).pipe(
+    mergeMap((action) =>
+      state$.pipe(
+        pluck('blockNumber'),
+        first((blockNumber) => action.payload.block <= blockNumber),
+        map(() => action),
+      ),
+    ),
+    withLatestFrom(from(userDepositContract.functions.balances(address))),
+    concatMap(([action, balance]) => {
+      try {
+        assert(balance.gt(Zero), [
+          ErrorCodes.UDC_PLAN_WITHDRAW_NO_BALANCE,
+          {
+            balance: balance.toString(),
+          },
+        ]);
+      } catch (e) {
+        return of(udcWithdraw.failure(e, { amount: action.meta.amount }));
+      }
+      const contract = getContractWithSigner(userDepositContract, signer);
+      return from(contract.functions.withdraw(action.meta.amount)).pipe(
```

Comment:
Don't use `from` inside `concatMap`, as it calls the function and creates the promise at creation time, instead of subscription time (which is the whole point of using concatMap instead of e.g. mergeMap)`. Use `defer(() => promise)` instead. Maybe the `balances` map & filter could/should also be under `defer` to ensure this evaluation happens when subscribed instead of before and to get an up-to-date view.